### Media ponderata

$=\frac{\sum_{i=1}^{n}x_{i}\cdot p_{i}}{\sum_{i=0}^{n} p_{i}}$

### Deviazione standard aka scarto quadratico medio

$A=\{3,6,4,5,3\}$  

Sia $n$ la popolosità totale.   
Sia $\mu_{A}$ la media di $A$.   
 

La deviazione standard di $A$ è data da;

$\sigma_{A}=\sqrt{\frac{\sum_{i=1}^{n}(A_{i}-\mu_{a})^{2}}{n}}=\sqrt{\frac{(A_{1}-\mu_{A})^2+(A_{...}-\mu_{A})^2+(A_{n}-\mu_{A})^2}{n}}$  

##### Svolgimento

$\mu_{A}=\frac{3+6+4+5+3}{5}=\frac{21}{5}=4.2$

$\sigma_{A}=\sqrt{\frac{(3-4.2)^2+(6-4.2)^2+(4-4.2)^2+(5-4.2)^2+(3-4.2)^2}{5}}=\sqrt{\frac{6.8}{5}}=\sqrt{1.36}=1.166$

In [28]:
#creazione di una funzione sigma di calcolo dello scarto quadratico medio in python3
def sigma(A):
    n = len(A)
    mi_a = sum(A) / n
    numerator = 0
    for a_i in A:
        numerator = numerator + ((a_i - mi_a) ** 2)
    result = (numerator / n) ** (1/2)
    return result

A = [3,6,4,5,3]

sigma(A)

1.16619037896906

### Varianza

$A=\{3,6,4,5,3\}$  

Sia $n$ la popolosità totale.   
Sia $\mu_{A}$ la media di $A$.   
 

La varianza di $A$ è data da;

$\sigma_{A}^2=\frac{\sum_{i=1}^{n}(A_{i}-\mu_{a})^{2}}{n}=\frac{(A_{1}-\mu_{A})^2+(A_{...}-\mu_{A})^2+(A_{n}-\mu_{A})^2}{n}$  
##### Svolgimento

$\mu_{A}=\frac{3+6+4+5+3}{5}=\frac{21}{5}=4.2$

$\sigma_{A}^2=\frac{(3-4.2)^2+(6-4.2)^2+(4-4.2)^2+(5-4.2)^2+(3-4.2)^2}{5}=\frac{6.8}{5}=1.36$


In [29]:
#creazione di una funzione sigma_square di calcolo della varianza in python3

def sigma_square(A):
    n = len(A)
    mi_a = sum(A) / n
    numerator = 0
    for a_i in A:
        numerator = numerator + ((a_i - mi_a) ** 2)
    result = numerator / n
    return result

A = [3,6,4,5,3]

sigma_square(A)

1.3599999999999999

 ### Covarianza

| Settimana | Lung omero | Lung femore | 
| --- | --- | --- | 
| 12 | 9 | 8 | 
| 16 | 20 | 20 | 
| 20 | 30 | 31 | 
| 24 | 40 | 42 | 
| 28 | 48 | 52 | 
| 32 | 55 | 61 | 
| 36 | 61 | 68 | 
| 40 | 66 | 74 | 

Sia $A$ la lunghezza dell' omero.  
Sia $B$ la lunghezza del femore.  
Sia $n$ la popolosità totale.  
Sia $\mu_{A}$ la media di $A$.  
Sia $\mu_{B}$ la media di $B$.  
La covarianza di $A$,$B$ è data da:  

$cov(A,B)=\frac{\sum_{i=1}^{n}(A_{i}-\mu_{A})(B_{i}-\mu_{B})}{n}=\frac{(A_{1}-\mu_{A})(B_{1}-\mu_{B})+(A_{...}-\mu_{A})(B_{...}-\mu_{B})+(A_{n}-\mu_{A})(B_{n}-\mu_{B})}{n}$  

##### Svolgimento

$\mu_{A}=\frac{9+20+30+40+48+55+61+66}{8}=\frac{329}{8}=41.125$  

$\mu_{B}=\frac{8+20+31+42+52+61+68+74}{8}=\frac{356}{8}=44.5$ 

$cov(A,B)=\frac{(9-41.125)(8-44.5)+(20-41.125)(20-44.5)+(30-41.125)(31-44.5)+(40-41.125)(42-44.5)+(48-41.125)(52-44.5)+(55-41.125)(61-44.5)+(61-41.125)(68-44.5)+(66-41.125)(74-44.5)}{8}=\frac{3324.5}{8}=415.562$

In [30]:
#creazione di una funzione cov di calcolo della covarianza in python3

def cov(A, B):
    n = len(A)
    mi_a = sum(A) / n
    mi_b = sum(B) / n
    numerator = 0
    for a_i, b_i in zip(A, B):
        numerator = numerator + (a_i - mi_a) * (b_i - mi_b)
    result = numerator / n
    return result

A = [9, 20, 30, 40, 48, 55, 61, 66]
B = [8, 20, 31, 42, 52, 61, 68, 74]

cov(A, B)

415.5625

#### Indice di correlazione di Pearson aka coefficiente di correlazione lineare di Pearson

| Settimana | Lung omero | Lung femore | 
| --- | --- | --- | 
| 12 | 9 | 8 | 
| 16 | 20 | 20 | 
| 20 | 30 | 31 | 
| 24 | 40 | 42 | 
| 28 | 48 | 52 | 
| 32 | 55 | 61 | 
| 36 | 61 | 68 | 
| 40 | 66 | 74 | 

Sia $A$ la lunghezza dell' omero.  
Sia $B$ la lunghezza del femore.  
Sia $cov(A,B)$ la varianza di $A,B$.  
Sia $\sigma_{A}$ la deviazione standard di $A$.    
Sia $\sigma_{B}$ la deviazione standard di $B$.    
L' indice di correlazione di Pearson tra $A$ e $B$ è dato da:

$\rho_{AB}=\frac{cov(A,B)}{\sigma_{A}\sigma_{B}}$


##### Svolgimento

$\mu_{A}=\frac{9+20+30+40+48+55+61+66}{8}=\frac{329}{8}=41.125$  

$\mu_{B}=\frac{8+20+31+42+52+61+68+74}{8}=\frac{356}{8}=44.5$ 

$cov(A,B)=\frac{(9-41.125)(8-44.5)+(20-41.125)(20-44.5)+(30-41.125)(31-44.5)+(40-41.125)(42-44.5)+(48-41.125)(52-44.5)+(55-41.125)(61-44.5)+(61-41.125)(68-44.5)+(66-41.125)(74-44.5)}{8}=\frac{3324.5}{8}=415.5625$

$\sigma_{A}=\sqrt{\frac{(9-41.125)^2+(20-41.125)^2+(30-41.125)^2+(40-41.125)^2+(48-41.125)^2+(55-41.125)^2+(61-41.125)^2+(66-41.125)^2}{8}}=\sqrt{\frac{2856.875}{8}}=\sqrt{375.109}=18.897$  

$\sigma_{B}=\sqrt{\frac{(8-44.5)^2+(20-44.5)^2+(31-44.5)^2+(42-44.5)^2+(52-44.5)^2+(61-44.5)^2+(68-44.5)^2+(74-44.5)^2}{8}}=\sqrt{\frac{3872}{8}}=\sqrt{484}=22$

$\rho_{AB}=\frac{415.562}{18.897\cdot22}=\frac{415.562}{415.734}=0.999$


In [31]:
"""
creazione di una funzione pearson_correlation_index di calcolo 
dell' indice di correlazione di Pearson in python3

!important verranno riutilizzate le funzioni precedentemente definite
"""

def pearson_correlation_index(A, B):
    cov_ab = cov(A, B)
    sigma_a = sigma(A)
    sigma_b = sigma(B)
    return cov_ab / (sigma_a * sigma_b)
    

A = [9, 20, 30, 40, 48, 55, 61, 66]
B = [8, 20, 31, 42, 52, 61, 68, 74]

pearson_correlation_index(A, B)

0.99956960974574

### Chi quadro  e indice V di Cramer

|  | Maschi | Femmine |
| --- | --- | --- | 
| Ripetenti | 12 | 4 |
| Non ripetenti | 7 | 16 |

Sia $t$ una generica cella.   
Sia $n$ la popolazione totale.  
Sia $r$ il numero di righe.  
Sia $c$ il numero di colonne.  
Sia $E[t_{rc}]$ il valore atteso della cella $t$ alla $r$esima riga e alla $c$esima colonna.  
Sia $\chi^2$ il Chi quadro.  
Sia $V$ l'indice di Cramer.

$E[t_{rc}]=\frac{(\sum_{i=0}^{r} t_{ic} )\cdot(\sum_{j=0}^{c} t_{rj} )}{n}$  

$\chi^2=\sum_{i=1}^{r}\sum_{j=1}^{c}\frac{(t_{ij} - E[t_{ij}])^2}{E[t_{ij}]}$  

$V=\sqrt{\frac{\frac{\chi^2}{n}}{min(r,c)-1}}$

##### Svolgimento
  
Tabella delle frequenze attese:  


|  | Maschi | Femmine | 
| --- | --- | --- | 
| Ripetenti | $\frac{16\cdot19}{39}$ | $\frac{16\cdot20}{39}$ |
| Non ripetenti | $\frac{23\cdot19}{39}$ | $\frac{23\cdot20}{39}$ |

|  |Maschi | Femmine | 
| --- | --- | --- | 
| Ripetenti | 7.794 | 8.205 |
| Non ripetenti | 11.205 | 11.794 |

$\chi^2 = \frac{(12-7.794)^2}{7.794} + \frac{(4-8.205)^2}{8.205} + \frac{(7-11.205)^2}{11.205} + \frac{(16-11.794)^2}{11.794}$  
$\chi^2 = \frac{17.690}{7.794} + \frac{17.682}{8.205} + \frac{17.682}{11.205} + \frac{17.690}{11.794}$  
$\chi^2 = 2.269 + 2.155 + 1.578 + 1.449$  
$\chi^2 = 7.451$  

$V=\sqrt{\frac{\frac{7.451}{39}}{1}} = \sqrt{\frac{7.451}{39}} = 0.09$

In [103]:
"""
creazione di una funzione chi_square di calcolo del chi quadro in python3
creazione di una funzione cramer_v_index di calcolo dell'indice di V cramer
"""

from functools import reduce
import numpy as np

def chi_square(A, e_A=None, n=None):
    rows_tots = list(map(lambda row: np.sum(row), A))
    cols_tots = list(map(lambda col: np.sum(col), A.T))
    if n is None:
        n = np.sum(A)
    e_A =  np.zeros_like(A, dtype=float)
    for i_cell, x in np.ndenumerate(e_A):
        r_tot = np.sum(A[i_cell[0]])
        c_tot = np.sum(A[:,i_cell[1]])
        e_A[i_cell] = r_tot * c_tot / n    
    res = 0
    for i_cell, x in np.ndenumerate(A):        
        e_x = e_A[i_cell]
        res += ((x - e_x) ** 2) / e_x
    return res

def cramer_v_index(chi_square, n, shape):
    return ((chi_square/n)/(min(shape) - 1)) ** 1/2
    
A = np.array([[12, 4], [7, 16]], dtype=float)
n = np.sum(A)
a_chi_square = chi_square(A, n=n)
a_v_cramer_index = cramer_v_index(a_chi_square, n=n, shape=A.shape)
print(a_chi_square, a_v_cramer_index)

7.5010297482837505 0.09616704805491988
